lib/srv/discovery/discovery.go

```diff
@@ -1327,8 +1322,78 @@ func (s *Server) getAllGCPServerFetchers() []server.Fetcher {
 	return allFetchers
 }
 
+// acquireDiscoveryGroup tries to acquire a lock if the Service has a DiscoveryGroup
+// It will retry using an exponential backoff algorithm.
+func (s *Server) acquireDiscoveryGroup() error {
+	if s.DiscoveryGroup == "" {
+		return nil
+	}
+
+	retry, err := retryutils.NewRetryV2(retryutils.RetryV2Config{
+		First:  0,
+		Driver: retryutils.NewExponentialDriver(defaults.HighResPollingPeriod),
+		Max:    defaults.LowResPollingPeriod,
+		Jitter: retryutils.NewHalfJitter(),
+		Clock:  s.clock,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var lease *services.SemaphoreLock
+	for range retry.After() {
+		retry.Inc()
+		s.Log.Debugf("Discovery service is trying to acquire lock for DiscoveryGroup %q", s.DiscoveryGroup)
+		lease, err = services.AcquireSemaphoreLock(s.ctx, services.SemaphoreLockConfig{
+			Service: s.Config.AccessPoint,
+			Expiry:  time.Minute,
+			Params: types.AcquireSemaphoreRequest{
+				SemaphoreKind: types.SemaphoreKindDiscoveryServiceGroup,
+				SemaphoreName: s.DiscoveryGroup,
+				MaxLeases:     1,
+				Holder:        s.ServerID,
+			},
+			Clock: s.clock,
+		})
+		if err == nil {
+			break
+		}
+
+		if !strings.Contains(err.Error(), teleport.MaxLeases) {
+			return trace.Wrap(err)
+		}
+		s.Log.Debugf("Discovery service is waiting on DiscoveryGroup %q lock: %v", s.DiscoveryGroup, err)
+	}
+
+	go func() {
+		for {
+			select {
+			case <-lease.Renewed():
+				continue
+			case <-lease.Done():
+				s.Log.WithError(lease.Wait()).Warnf("DiscoveryGroup %q lock was lost, stopping discovery service", s.DiscoveryGroup)
+				s.Stop()
+				return
+			}
+		}
+	}()
+
+	s.releaseGroupLockFn = func() {
```

Comment:
Do we need to synchronize the release function, from a memory model perspective and from a logical one? What if we end up stopping the `*Server` before we managed to update the `releaseGroupLockFn`?

Reply:
I moved the `releaseGroupLockFn` creation to much earlier.
Now, it's the first thing it does when calling the `Start`

This should ensure it's very unlikely to call the `Stop` before `releaseGroupLockFn` was set.

Reply:
"Very unlikely" doesn't mean anything; we should set a flag in Stop (while holding the lock) that means "server has been stopped" and we need to check the flag (while holding the lock) as we set `releaseGroupLockFn`, and we should refuse to start if the flag is set (and thus nothing can ever stop the server again).

Reply:
Can you please check again?
I've simplified the flow and I'm no longer using the `releaseGroupLockFn`
The AcquireSemaphoreLock receives the Server's Context and should end as soon as we cancel that context.